<a href="https://colab.research.google.com/github/AnkitWalde/EDA_on_Hotel_Booking.ipynb/blob/main/EDA_on_Hotel_Booking_Analysis_Capstone_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## <b> Have you ever wondered when the best time of year to book a hotel room is? Or the optimal length of stay in order to get the best daily rate? What if you wanted to predict whether or not a hotel was likely to receive a disproportionately high number of special requests? This hotel booking dataset can help you explore those questions!

## <b>This data set contains booking information for a city hotel and a resort hotel, and includes information such as when the booking was made, length of stay, the number of adults, children, and/or babies, and the number of available parking spaces, among other things. All personally identifying information has been removed from the data. </b>

## <b> Explore and analyze the data to discover important factors that govern the bookings. </b>

 **IMPORT LIBRARIES**

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

**MOUNT DRIVE**

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


**SET FILE PATH**

In [3]:
file_path = '/content/drive/MyDrive/Almabetter/Capstone EDA Project/Explolatory Data Analysis on Hotel Bookings/Hotel Bookings.csv'

In [4]:
df = pd.read_csv(file_path)

* Now that we have the required data,I will take my first step to become a data scientist. WISH ME LUCK!!

* First,let's make a copy of our data so our original file 
stay intact.

In [5]:
df_hotel = df.copy()

* We will have an overview of the data in the file.

In [6]:
df_hotel.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  deposit_type  \
0                        0                     0       2  ...    No Deposit   
1                        0                     0       2  ...    No Deposit   
2                        0                     1       1  ...    No Deposit   
3                        0                     1       1  ...    No Deposit   
4                        0                     2       2  ...    No Deposit   

   agent company days_in_waiting_list customer_type   adr  \
0    NaN     NaN                    0     Transient   0.0   
1    NaN     NaN                    0     Transient   0.0   
2    NaN     NaN                    0     Transient  75.0   
3  304.0     NaN                    0     Transient  75.0   
4  240.0     NaN                    0     Transient  98.0   

   required_car_parking_spaces  total_of_special_requests  reservation_status  \
0                            0                          0           Check-Out   
1                            0                          0           Check-Out   
2                            0                          0           Check-Out   
3                            0                          0           Check-Out   
4                            0                          1           Check-Out   

  reservation_status_date  
0              2015-07-01  
1              2015-07-01  
2              2015-07-02  
3              2015-07-02  
4              2015-07-03  

[5 rows x 32 columns]

In [7]:
df_hotel.tail()

hotel  is_canceled  lead_time  arrival_date_year  \
119385  City Hotel            0         23               2017   
119386  City Hotel            0        102               2017   
119387  City Hotel            0         34               2017   
119388  City Hotel            0        109               2017   
119389  City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
119385             August                        35   
119386             August                        35   
119387             August                        35   
119388             August                        35   
119389             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  adults  ...  deposit_type  agent company  \
119385                     5       2  ...    No Deposit  394.0     NaN   
119386                     5       3  ...    No Deposit    9.0     NaN   
119387                     5       2  ...    No Deposit    9.0     NaN   
119388                     5       2  ...    No Deposit   89.0     NaN   
119389                     7       2  ...    No Deposit    9.0     NaN   

       days_in_waiting_list customer_type     adr  \
119385                    0     Transient   96.14   
119386                    0     Transient  225.43   
119387                    0     Transient  157.71   
119388                    0     Transient  104.40   
119389                    0     Transient  151.20   

        required_car_parking_spaces  total_of_special_requests  \
119385                            0                          0   
119386                            0                          2   
119387                            0                          4   
119388                            0                          0   
119389                            0                          2   

        reservation_status reservation_status_date  
119385           Check-Out              2017-09-06  
119386           Check-Out              2017-09-07  
119387           Check-Out              2017-09-07  
119388           Check-Out              2017-09-07  
119389           Check-Out              2017-09-07  

[5 rows x 32 columns]

* Now we will see the data types of entries and null values in our columns.

In [8]:
df_hotel.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

* Here we can see that there are a total of 
119390 entries,in which there are 32 columns in the dataframe and columns like 'children','company','country' and 'agent' have null values.

* Next,we will do a quick check on the given data.




In [9]:
df_hotel.describe(include='all')

hotel    is_canceled      lead_time  arrival_date_year  \
count       119390  119390.000000  119390.000000      119390.000000   
unique           2            NaN            NaN                NaN   
top     City Hotel            NaN            NaN                NaN   
freq         79330            NaN            NaN                NaN   
mean           NaN       0.370416     104.011416        2016.156554   
std            NaN       0.482918     106.863097           0.707476   
min            NaN       0.000000       0.000000        2015.000000   
25%            NaN       0.000000      18.000000        2016.000000   
50%            NaN       0.000000      69.000000        2016.000000   
75%            NaN       1.000000     160.000000        2017.000000   
max            NaN       1.000000     737.000000        2017.000000   

       arrival_date_month  arrival_date_week_number  \
count              119390             119390.000000   
unique                 12                       NaN   
top                August                       NaN   
freq                13877                       NaN   
mean                  NaN                 27.165173   
std                   NaN                 13.605138   
min                   NaN                  1.000000   
25%                   NaN                 16.000000   
50%                   NaN                 28.000000   
75%                   NaN                 38.000000   
max                   NaN                 53.000000   

        arrival_date_day_of_month  stays_in_weekend_nights  \
count               119390.000000            119390.000000   
unique                        NaN                      NaN   
top                           NaN                      NaN   
freq                          NaN                      NaN   
mean                    15.798241                 0.927599   
std                      8.780829                 0.998613   
min                      1.000000                 0.000000   
25%                      8.000000                 0.000000   
50%                     16.000000                 1.000000   
75%                     23.000000                 2.000000   
max                     31.000000                19.000000   

        stays_in_week_nights         adults  ...  deposit_type          agent  \
count          119390.000000  119390.000000  ...        119390  103050.000000   
unique                   NaN            NaN  ...             3            NaN   
top                      NaN            NaN  ...    No Deposit            NaN   
freq                     NaN            NaN  ...        104641            NaN   
mean                2.500302       1.856403  ...           NaN      86.693382   
std                 1.908286       0.579261  ...           NaN     110.774548   
min                 0.000000       0.000000  ...           NaN       1.000000   
25%                 1.000000       2.000000  ...           NaN       9.000000   
50%                 2.000000       2.000000  ...           NaN      14.000000   
75%                 3.000000       2.000000  ...           NaN     229.000000   
max                50.000000      55.000000  ...           NaN     535.000000   

            company days_in_waiting_list customer_type            adr  \
count   6797.000000        119390.000000        119390  119390.000000   
unique          NaN                  NaN             4            NaN   
top             NaN                  NaN     Transient            NaN   
freq            NaN                  NaN         89613            NaN   
mean     189.266735             2.321149           NaN     101.831122   
std      131.655015            17.594721           NaN      50.535790   
min        6.000000             0.000000           NaN      -6.380000   
25%       62.000000             0.000000           NaN      69.290000   
50%      179.000000             0.000000           NaN      94.575000   
75%      270.000000             0.000000           

**DATA DESCRIPTION**
- **hotel** : Name of hotel ( City or Resort)
- **is_canceled** : Whether the booking is canceled or not (0 for no canceled and 1 for canceled)
- **lead_time** : time (in days) between booking transaction and actual arrival.
- **arrival_date_year** : Year of arrival
- **arrival_date_month** : month of arrival
- **arrival_date_week_number** : week number of arrival date.
- **arrival_date_day_of_month** : Day of month of arrival date
- **stays_in_weekend_nights** : No. of weekend nights spent in a hotel
- **stays_in_week_nights** : No. of weeknights spent in a hotel
- **adults** : No. of adults in single booking record.
- **children** : No. of children in single booking record.
- **babies** : No. of babies in single booking record. 
- **meal** : Type of meal chosen 
- **country** : Country of origin of customers (as mentioned by them)
- **market_segment** : What segment via booking was made and for what purpose.
- **distribution_channel** : Via which medium booking was made.
- **is_repeated_guest** : Whether the customer has made any booking before(0 for No and 1 for yes)                  
- **previous_cancellations** : No. of previous canceled bookings.
- **previous_bookings_not_canceled** : No. of previous non-canceled bookings.
- **reserved_room_type** : Room type reserved by a customer.
- **assigned_room_type** : Room type assigned to the customer.
- **booking_changes** : No. of booking changes done by customers
- **deposit_type** : Type of deposit at the time of making a booking (No deposit/ Refundable/ No refund)
- **agent** : Id of agent for booking
- **company** : Id of the company making a booking
- **days_in_waiting_list** : No. of days on waiting list.
- **customer_type** : Type of customer(Transient, Group, etc.)
- **adr** : Average Daily rate.
- **required_car_parking_spaces** : No. of car parking asked in booking
- **total_of_special_requests** : total no. of special request.
- **reservation_status** : Whether a customer has checked out or canceled,or not showed 
- **reservation_status_date** : Date of making reservation status.







In [10]:
# columns name list
print(list(df_hotel))

['hotel', 'is_canceled', 'lead_time', 'arrival_date_year', 'arrival_date_month', 'arrival_date_week_number', 'arrival_date_day_of_month', 'stays_in_weekend_nights', 'stays_in_week_nights', 'adults', 'children', 'babies', 'meal', 'country', 'market_segment', 'distribution_channel', 'is_repeated_guest', 'previous_cancellations', 'previous_bookings_not_canceled', 'reserved_room_type', 'assigned_room_type', 'booking_changes', 'deposit_type', 'agent', 'company', 'days_in_waiting_list', 'customer_type', 'adr', 'required_car_parking_spaces', 'total_of_special_requests', 'reservation_status', 'reservation_status_date']


* Finding unique values to better understand all the columns of the dataframe.

In [11]:
df_hotel['hotel'].unique()

# Also we can Find unique values in dataset using this code 
# df_new['column_name'].unique()
# but this method won't work on large no of columns

array(['Resort Hotel', 'City Hotel'], dtype=object)

In [12]:
#another method for finding unique values
column_name=list(set(df_hotel.drop(columns=['reservation_status_date','country','arrival_date_month']).columns)-set(df_hotel.describe()))
for i in column_name:
  print(f'Unique values in column {i} are:, {(df_hotel[i].unique())}')

Unique values in column reservation_status are:, ['Check-Out' 'Canceled' 'No-Show']
Unique values in column market_segment are:, ['Direct' 'Corporate' 'Online TA' 'Offline TA/TO' 'Complementary' 'Groups'
 'Undefined' 'Aviation']
Unique values in column reserved_room_type are:, ['C' 'A' 'D' 'E' 'G' 'F' 'H' 'L' 'P' 'B']
Unique values in column distribution_channel are:, ['Direct' 'Corporate' 'TA/TO' 'Undefined' 'GDS']
Unique values in column customer_type are:, ['Transient' 'Contract' 'Transient-Party' 'Group']
Unique values in column assigned_room_type are:, ['C' 'A' 'D' 'E' 'G' 'F' 'I' 'B' 'H' 'P' 'L' 'K']
Unique values in column hotel are:, ['Resort Hotel' 'City Hotel']
Unique values in column meal are:, ['BB' 'FB' 'HB' 'SC' 'Undefined']
Unique values in column deposit_type are:, ['No Deposit' 'Refundable' 'Non Refund']


*   There are a lot of duplicate data which  can affect the outcome of EDA needs to be deleted. 
*   Rows and column should be shaped .
*   Removing the Null Values from the data.

*  New important column for better analysing should be created.
*  Conversion of column into  the new data type according to need.

##Data Cleaning

In [13]:
# finding the duplicate rows
# True means duplicate rows
df_hotel.duplicated().value_counts()

False    87396
True     31994
dtype: int64

In [14]:
#dropping the duplicate rows
df_hotel = df_hotel.drop_duplicates()


In [15]:
#checking the column was drop or not
df_hotel.shape

(87396, 32)

In [16]:
df_hotel.isnull().sum().sort_values(ascending = False)[:8].reset_index()

index      0
0             company  82137
1               agent  12193
2             country    452
3            children      4
4  reserved_room_type      0
5  assigned_room_type      0
6     booking_changes      0
7        deposit_type      0

In [17]:
#filling null value in company,agent with 0
df_hotel[['company','agent']] = df_hotel[['company','agent']].fillna(0)

#filling country null value with others
new_columns = ['country']
for i in new_columns:
  df_hotel[i].fillna(0,inplace = True)

#filling children null value with mean
df_hotel['children'].fillna(df_hotel['children'].mean(),inplace = True)

In [18]:
#checking null values is change or not
df_hotel.isnull().sum().sort_values(ascending = False)[:8].reset_index()

index  0
0                        hotel  0
1                  is_canceled  0
2           reservation_status  0
3    total_of_special_requests  0
4  required_car_parking_spaces  0
5                          adr  0
6                customer_type  0
7         days_in_waiting_list  0

*There are some rows (babies,children or adults) is equal to zero. So we will remove this 

In [19]:
#checking the total number of rows which containing value equal to 0
df_hotel[df_hotel['babies']+df_hotel['children']+df_hotel['adults'] == 0].shape

(166, 32)

In [20]:
#dropping those rows
df_hotel.drop(df_hotel[df_hotel['babies']+df_hotel['children']+df_hotel['adults'] == 0].index, inplace = True)

In [21]:
#add some new columns
df_hotel['total_people'] = df_hotel['babies']+df_hotel['children']+df_hotel['adults']
df_hotel['total_stay'] = df_hotel['stays_in_weekend_nights'] + df_hotel['stays_in_week_nights']

* By adding this we can analyse better the total stay length of hotel

In [22]:
#checking the new columns has added
df_hotel.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  company  \
0                        0                     0       2  ...      0.0   
1                        0                     0       2  ...      0.0   
2                        0                     1       1  ...      0.0   
3                        0                     1       1  ...      0.0   
4                        0                     2       2  ...      0.0   

   days_in_waiting_list customer_type   adr required_car_parking_spaces  \
0                     0     Transient   0.0                           0   
1                     0     Transient   0.0                           0   
2                     0     Transient  75.0                           0   
3                     0     Transient  75.0                           0   
4                     0     Transient  98.0                           0   

  total_of_special_requests  reservation_status  reservation_status_date  \
0                         0           Check-Out               2015-07-01   
1                         0           Check-Out               2015-07-01   
2                         0           Check-Out               2015-07-02   
3                         0           Check-Out               2015-07-02   
4                         1           Check-Out               2015-07-03   

   total_people total_stay  
0           2.0          0  
1           2.0          0  
2           1.0          1  
3           1.0          1  
4           2.0          2  

[5 rows x 34 columns]

In [23]:
#checking the final shape of rows and columns
df_hotel.shape

(87230, 34)

##**Exploratory Data Analysis**